File: Sources/windows/BitTherapyCLR/src/game/game.cpp
#include "game.h"

#include <chrono>
#include <iostream>
#include <numbers>
#include <sstream>

#include "geometry.h"
#include "../utils/utils.h"

RenderedItem::RenderedItem(
    uint32_t id, 
    std::string spritePath, 
    Rect frame,
    bool isFlipped,
    double zRotation
) : 
    id(id),
    spritePath(spritePath), 
    frame(frame),
    isFlipped(isFlipped),
    zRotation(zRotation)
{}

Game::Game(std::string screenName, Rect bounds) : 
    screenName(screenName),
    bounds(bounds),
    entities(std::vector<Entity*>({})) 
{}

void Game::update(std::chrono::milliseconds timeSinceLastUpdate) {
    std::lock_guard<std::mutex> lock(mtx);
    for (auto& entity : entities) {
        entity->update(timeSinceLastUpdate);
    }
}

void Game::addEntities(const std::vector<Entity*> entities) {
    for (Entity* entity : entities) {
        addEntity(entity);
    }
}

void Game::addEntity(Entity* entity) {
    std::lock_guard<std::mutex> lock(mtx);
    entities.push_back(entity);
}

const uint32_t Game::numberOfEntities() {
    std::lock_guard<std::mutex> lock(mtx);
    return entities.size();
}

std::vector<RenderedItem> Game::render() {
    std::lock_guard<std::mutex> lock(mtx);
    std::vector<RenderedItem> renderedItems({});
    
    for (const auto& entity : entities) {
        auto item = RenderedItem(
            entity->id,
            entity->currentSpriteFrame(), 
            entity->frame,
            isFlipped(entity),
            rotation(entity)
        );
        renderedItems.push_back(item);
    }
    return renderedItems;
}

double Game::isFlipped(const Entity* entity) {
    return entity->direction.x < 0.0;
}

double Game::rotation(const Entity* entity) {
    if (entity->direction.x == 0.0 && entity->direction.y == 0.0) {
        return 0.0;
    }
    if (entity->direction.x == 0.0 && entity->direction.y < 0.0) {
        return 1.5 * std::numbers::pi;
    }
    if (entity->direction.x == 0.0 && entity->direction.y > 0.0) {
        return 0.5 * std::numbers::pi;
    }
    if (entity->direction.x <= 0.0 && entity->direction.y == 0.0 && entity->frame.y == bounds.y) {
        return 1.0 * std::numbers::pi;
    }
    return 0.0;
}
    
void Game::mouseDragStarted(const uint32_t& targetId) {
    std::lock_guard<std::mutex> lock(mtx);
    for (auto& entity : entities) {
        if (entity->id == targetId) {
            auto dragSprite = entity->species->dragPath;
            entity->direction = Vector2d(0.0, 0.0);
            entity->changeSprite(dragSprite);
        }
    }    
}

void Game::mouseDragEnded(const uint32_t& targetId, const Vector2d& delta) {
    std::lock_guard<std::mutex> lock(mtx);
    for (auto& entity : entities) {
        if (entity->id == targetId) {
            auto dx = delta.x > 0 ? 1.0 : -1.0;
            auto movementSprite = entity->species->movementPath;

            entity->frame = entity->frame.offset(delta);
            entity->frame.x = fmax(entity->frame.x, 0.0);
            entity->frame.x = fmin(entity->frame.x, bounds.w - entity->frame.w);
            entity->frame.y = fmax(entity->frame.y, 0.0);
            entity->frame.y = fmin(entity->frame.y, bounds.h - entity->frame.h);
            entity->direction = Vector2d(dx, 0.0);
            entity->changeSprite(movementSprite);
        }
    }    
}

std::string Game::description() {
    std::lock_guard<std::mutex> lock(mtx);
    std::stringstream ss; 

    ss << "Game @ " << this << std::endl;
    ss << "  Screen: " << screenName << std::endl;
    ss << "  Origin: " << bounds.x << ", " << bounds.y << std::endl;
    ss << "  Size: " << bounds.w << " x " << bounds.h << std::endl;
    ss << "  Entities: " << entities.size() << std::endl << std::endl;

    for (const auto& entity : entities) {
        ss << entity->description() << std::endl;
    }
    return ss.str();
}

File: Sources/windows/App/Sources/Game/Game.cs
using App;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public class Game
    {
        private readonly object _lock = new object();
        private readonly List<Entity> entities = new List<Entity>();

        public string ScreenName { get; }
        public Rect Bounds { get; }

        public Game(string screenName, Rect bounds)
        {
            ScreenName = screenName;
            Bounds = bounds;
        }

        public void Update(TimeSpan timeSinceLastUpdate)
        {
            lock (_lock)
            {
                foreach (var entity in entities)
                {
                    entity.Update(timeSinceLastUpdate);
                }
            }
        }

        public void AddEntity(Entity entity)
        {
            lock (_lock)
            {
                entities.Add(entity);
            }
        }

        public void AddEntities(IEnumerable<Entity> entities)
        {
            lock (_lock)
            {
                this.entities.AddRange(entities);
            }
        }

        public uint NumberOfEntities()
        {
            lock (_lock)
            {
                return (uint)entities.Count;
            }
        }

        public List<RenderedItem> Render()
        {
            lock (_lock)
            {
                var renderedItems = new List<RenderedItem>();
                foreach (var entity in entities)
                {
                    var item = new RenderedItem(
                        entity.Id,
                        entity.CurrentSpriteFrame(),
                        entity.Frame,
                        IsFlipped(entity),
                        Rotation(entity)
                    );
                    renderedItems.Add(item);
                }
                return renderedItems;
            }
        }

        private bool IsFlipped(Entity entity)
        {
            return entity.Direction.X < 0.0;
        }

        private double Rotation(Entity entity)
        {
            if (entity.Direction.X == 0.0 && entity.Direction.Y == 0.0)
            {
                return 0.0;
            }
            if (entity.Direction.X == 0.0 && entity.Direction.Y < 0.0)
            {
                return 1.5 * Math.PI;
            }
            if (entity.Direction.X == 0.0 && entity.Direction.Y > 0.0)
            {
                return 0.5 * Math.PI;
            }
            if (entity.Direction.X <= 0.0 && entity.Direction.Y == 0.0 && entity.Frame.Y == Bounds.Y)
            {
                return Math.PI;
            }
            return 0.0;
        }

        public void MouseDragStarted(uint targetId)
        {
            lock (_lock)
            {
                foreach (var entity in entities)
                {
                    if (entity.Id == targetId)
                    {
                        var dragSprite = entity.Species.DragPath;
                        entity.Direction = new Vector2d(0.0, 0.0);
                        entity.ChangeSprite(dragSprite);
                    }
                }
            }
        }

        public void MouseDragEnded(uint targetId, float deltaX, float deltaY)
        {
            lock (_lock)
            {
                foreach (var entity in entities)
                {
                    if (entity.Id == targetId)
                    {
                        var dx = deltaX > 0 ? 1.0 : -1.0;
                        var movementSprite = entity.Species.MovementPath;

                        var newFrame = entity.Frame.Offset(deltaX, deltaY);
                        newFrame.X = Math.Max(entity.Frame.X, 0.0);
                        newFrame.X = Math.Min(entity.Frame.X, Bounds.W - entity.Frame.W);
                        newFrame.Y = Math.Max(entity.Frame.Y, 0.0);
                        newFrame.Y = Math.Min(entity.Frame.Y, Bounds.H - entity.Frame.H);
                        entity.Frame = newFrame;
                        entity.Direction = new Vector2d(dx, 0.0);
                        entity.ChangeSprite(movementSprite);
                    }
                }
            }
        }

        public string Description()
        {
            lock (_lock)
            {
                var sb = new StringBuilder();

                sb.AppendLine($"Game @ {this}");
                sb.AppendLine($"  Screen: {ScreenName}");
                sb.AppendLine($"  Origin: {Bounds.X}, {Bounds.Y}");
                sb.AppendLine($"  Size: {Bounds.W} x {Bounds.H}");
                sb.AppendLine($"  Entities: {entities.Count}");
                sb.AppendLine();

                foreach (var entity in entities)
                {
                    sb.AppendLine(entity.Description());
                }
                return sb.ToString();
            }
        }
    }
}
